Tools/cases_generator/optimizer_generator.py

```diff
@@ -147,16 +163,98 @@ def goto_label(self, goto: Token, label: Token, storage: Storage) -> None:
         self.out.emit(goto)
         self.out.emit(label)
 
+    def replace_opcode_if_evaluates_pure(
+        self,
+        tkn: Token,
+        tkn_iter: TokenIterator,
+        uop: CodeSection,
+        storage: Storage,
+        inst: Instruction | None,
+    ) -> bool:
+        skip_to(tkn_iter, "SEMI")
+        return True
+
+class OptimizerConstantEmitter(OptimizerEmitter):
+    def __init__(self, out: CWriter, labels: dict[str, Label], uop: Uop):
+        super().__init__(out, labels)
+        # Replace all outputs to point to their stackref versions.
+        overrides = {
+            outp.name: self.emit_stackref_override for outp in uop.stack.outputs
+        }
+        self._replacers = {**self._replacers, **overrides}
+
+    def emit_stackref_override(
+        self,
+        tkn: Token,
+        tkn_iter: TokenIterator,
+        uop: CodeSection,
+        storage: Storage,
+        inst: Instruction | None,
+    ) -> bool:
+        self.out.emit(tkn)
+        self.out.emit("_stackref ")
+        return True
+
+def write_uop_pure_evaluation_region_header(
+    uop: Uop,
+    out: CWriter,
+    stack: Stack,
+) -> None:
+    emitter = OptimizerConstantEmitter(out, {}, uop)
+    emitter.emit("if (\n")
+    assert len(uop.stack.inputs) > 0, "Pure operations must have at least 1 input"
+    for inp in uop.stack.inputs[:-1]:
```

Comment:
Note: this should take the inputs from `REPLACE_OUTPUT_IF_EVALUATES_PURE`, not all inputs but I will do that in a follow-up PR. I want to land this because the diff is becoming too big and drifting from main soon. Please excuse this for now.